kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+  -- initiates an asynchronous background updating task to refresh it.
+  local ttl, _, answers = self.cache:peek(key, true)
+  if answers and not answers.errcode and ttl then
+    if not answers.expired then
+      answers.expire = now() + ttl
+      answers.expired = true
+      ttl = ttl + self.stale_ttl
+
+    else
+      ttl = ttl + (answers.expire - now())
+    end
+
+    -- trigger the update task by the upper caller every 60 seconds
+    ttl = math_min(ttl, 60)
+
+    if ttl > 0 then
+      log(DEBUG, PREFIX, "start stale update task ", key, " ttl:", ttl)
+
+      -- mlcache's internal lock mechanism ensures concurrent control
+      start_stale_update_task(self, key, name, qtype)
+      answers.ttl = ttl
+      return answers, nil, ttl
+    end
+  end
+
+  if cache_only then
+    return CACHE_ONLY_ANSWERS, nil, -1
+  end
+
+  log(DEBUG, PREFIX, "cache miss, try to query ", key)
+
+  return resolve_query_types(self, name, qtype, tries)
+end
+
+
+local function resolve_all(self, name, qtype, cache_only, tries, has_timing)
+  name = string_lower(name)
+  tries = setmetatable(tries or {}, TRIES_MT)
+
+  if not qtype then
+    qtype = ((self.enable_srv and is_srv(name)) and TYPE_SRV or TYPE_A_AAAA)
+  end
+
+  local key = name .. ":" .. qtype
+
+  log(DEBUG, PREFIX, "resolve_all ", key)
+
+  stats_init_name(self.stats, key)
+  stats_increment(self.stats, key, "runs")
+
+  local answers, err, hit_level = self.cache:get(key, nil, resolve_callback,
+                                                 self, name, qtype, cache_only,
+                                                 tries)
+  -- check for runtime errors in the callback
+  if err and err:sub(1, 8) == "callback" then
+    log(ALERT, PREFIX, err)
+  end
+
+  local hit_str = hit_level and HIT_LEVEL_TO_NAME[hit_level] or "fail"
+  stats_increment(self.stats, key, hit_str)
+
+  log(DEBUG, PREFIX, "cache lookup ", key, " ans:", answers and #answers or "-",
+             " hlv:", hit_str)
+
+  if has_timing then
+    req_dyn_hook_run_hook("timing", "dns:cache_lookup",
+                          (hit_level and hit_level < HIT_L3))
+  end
+
+  if answers and answers.errcode then
+    err = ("dns %s error: %s %s"):format(
+            answers.errcode < CACHE_ONLY_ERROR_CODE and "server" or "client",
+            answers.errcode, answers.errstr)
+    return nil, err, tries
+  end
+
+  return answers, err, tries
+end
+
+
+function _M:resolve(name, qtype, cache_only, tries)
+  return resolve_all(self, name, qtype, cache_only, tries,
+                     ngx.ctx and ngx.ctx.has_timing)
+end
+
+
+function _M:resolve_address(name, port, cache_only, tries)
+  local has_timing = ngx.ctx and ngx.ctx.has_timing
+
+  local answers, err, tries = resolve_all(self, name, nil, cache_only, tries,
+                                          has_timing)
+  if not answers then
+    return nil, err, tries
+  end
+
+  if answers and answers[1].type == TYPE_SRV then
```

Comment:
Could `answers[1]` be a nil?

Reply:
it could not be nil, but I think for code understanding, we should check it. and the previous logic has checked the answers, so here we just need to check ` if answers[1] and answers[1].type == TYPE_SRV then`

Reply:
done, and make the code simpler